geopandas/clip.py

```diff
@@ -0,0 +1,246 @@
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+          Polygon vector layer used to clip gdf.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with gdf.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from gdf clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global capital cities) with a polygon (the South American continent):
+
+        >>> import geopandas as gpd
+        >>> import geopandas.clip as gc
+        >>> path = geopandas.datasets.get_path('naturalearth_lowres')
+        >>> world = geopandas.read_file(path)
+        >>> south_america = world[world['continent'] == "South America"]
+        >>> capitals = gpd.read_file(gpd.datasets.get_path('naturalearth_cities'))
+        >>> capitals.shape
+        (202, 2)
+        >>> sa_capitals = cl.clip(capitals, south_america)
+        >>> sa_capitals.shape
+        (12, 2)
+    """
+    if not isinstance(gdf, (gpd.GeoDataFrame, gpd.GeoSeries)):
+        raise AttributeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(
+        clip_obj, (gpd.GeoDataFrame, gpd.GeoSeries, Polygon, MultiPolygon)
+    ):
+        raise AttributeError(
+            "'clip_obj' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(clip_obj, (gpd.GeoDataFrame, gpd.GeoSeries)):
+        xmin, ymin, xmax, ymax = clip_obj.total_bounds
+    else:
+        xmin, ymin, xmax, ymax = clip_obj.bounds
+    if gdf.cx[xmin:xmax, ymin:ymax].empty:
+        raise ValueError("Shape and crop extent do not overlap.")
+
+    poly = clip_obj.geometry.unary_union
+
+    geom_types = gdf.geometry.type
+    line_idx = np.asarray((geom_types == "Polygon") | (geom_types == "LineString"))
+    multiline_idx = np.asarray(
+        (geom_types == "MultiPolygon") | (geom_types == "MultiLineString")
+    )
+    point_idx = np.asarray((geom_types == "Point"))
+    multipoint_idx = np.asarray((geom_types == "MultiPoint"))
+
+    multipoints = gdf[multipoint_idx]
+    if not multipoints.empty:
+        multipoint_gdf = _clip_multi_point(multipoints, poly)
+    else:
+        multipoint_gdf = None
+    points = gdf[point_idx]
+    if not points.empty:
+        point_gdf = _clip_points(points, poly)
+    else:
+        point_gdf = None
+    multilines = gdf[multiline_idx]
+    if not multilines.empty:
+        multiline_gdf = _clip_multi_poly_line(multilines, poly)
+    else:
+        multiline_gdf = None
+    lines = gdf[line_idx]
+    if not lines.empty:
+        line_gdf = _clip_line_poly(lines, poly)
+    else:
+        line_gdf = None
+
+    return pd.concat([multipoint_gdf, point_gdf, multiline_gdf, line_gdf])
```

Comment:
@martinfleis one "problem" with this is that it changes the order of the rows in the dataframe (which might be a bit unexpected for a clip operation?) 
We could do the effort to re-order again to match the original order, if desired (only in the case it is needed of course)

Reply:
That is a good point. Something along these lines? @lwasser can you add something like this?

```py
order = pd.Series(range(len(gdf)), index=gdf.index)
concat = pd.concat([multipoint_gdf, point_gdf, multiline_gdf, line_gdf])
concat['_order'] = order
return concat.sort_values(by='_order').drop(columns='_order')
```

Reply:
we are playing with that now @martinfleis  and when i say we i mean @nkorinek is working on it :) thank you again!!

Reply:
ok this fix was just submitted.